### DISCUSSION TITLE
Fetch composable and Pinia - nuxt instance unavailable

### DISCUSSION THREAD

**Discussion by milelazar** [nuxt/nuxt#17368] [Questions] [answered]
On first app load, I need to fetch some initial data from API endpoints that require authorization token and save the data to Pinia. So to avoid repeating headers for each $fetch, I've created a $fetch wrapper composable which is used within store actions.

And this works fine if composable is used only once. If used more than once, I get "nuxt instance unavailable" error.
Not sure what am I doing wrong.

Nuxt version: 3.0.0-rc.4

Here is the example:

store/cms.js
```
import { defineStore } from 'pinia'
export const cmsStore = defineStore({
	id: 'cms',
	state: () => {
		return {
			endpoints: [],
			menus: [],
		}
	},
	actions: {
		async fetchData() {
			const config = useConfig()

			// get endpoints
			const endpoints = await useApi(`${config.apiUrl}/misc/endpoints?version=v1`).then(res => {
				this.endpoints = endpoints;
				return res.data;
			});
		
			// get menus
			await useApi(`${config.apiUrl}${endpoints._get_hapi_cms_menus'}`).then(res => this.menus = res.data);
	}
})
```
If "get menus" fetch is removed or replaced with default $fetch, everything works.

composables/useApi.js
```
export const useApi = async (url, options) => {
	const token = useState('hapi_auth_token');
	return $fetch(url, {
		headers: {
			'Authorization': 'Bearer ' + token.value
		},
		options
	})
}
```
If hardcoded token is used instead of useState, everything works.

plugins/index.js
```
import { cmsStore } from "@/store/cms";
export default defineNuxtPlugin(async (nuxtApp) => {
	// api token is generated here and passed to useState
	useState('hapi_auth_token', () => 'api_auth_token');

	const store = cmsStore()
	await store.fetchData()
})
```

pages/[...slug].vue
```
<template>
	<div>Page</div>
</template>
```

**Comment by Diizzayy** [comment, marked as answer]
@milelazar I suspect that this is caused by the multiple await calls to your `useApi` composable which utilizes `useState`, note that composables aren't meant to house multiple await calls in that manner, especially if a `Nuxt composable` is called anywhere after an await ( hence the `nuxt instance unavailable` error).

Here is an example of how your code can be restructured:

```ts
async fetchData() {
  const config = useConfig()

  await Promise.all([
    // get endpoints
    useApi(`${config.apiUrl}/misc/endpoints?version=v1`),
    // get menus
    useApi(`${config.apiUrl}${endpoints._get_hapi_cms_menus}`)
  ]).then(([endpoints, menus]) => {
    this.endpoints = endpoints.data
    this.menus = menus.data
  })
}
```

**Reply by milelazar** [reply]
@Diizzayy it makes sense. Works fine now. Thank you for the explanation and solution!
